### PR TITLE
Rotation Matrix Implementation

### PR DESCRIPTION
**Pull request by viksharma-mtv**
#### Changes since PR #3

- Change to TYPE_ROTATION_VECTOR (fixed deprecation issue)
- Remapping of the coordinate system along flipped X-axis and standard Y-axis
- Passing in the remapped rotation matrix directly to the MediaPipe graph and setting the transpose (inverse) as the IMU rotation submatrix
- Basic normals calculations (This is being updated to newer shading system in PR #5)

### CODE REVIEW

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 1]
```diff
@@ -626,36 +635,45 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
...
+    varying mediump vec3 sample_coordinate;
+    varying mediump vec3 vNormal;
 
     void main() {
-      sample_coordinate = texture_coordinate.xy;
+      sample_coordinate = texture_coordinate.xyz;
       mat4 mvpMatrix = perspectiveMatrix * modelMatrix;
       gl_Position = mvpMatrix * position;
+
+      vec4 tmpNormal = mvpMatrix * vec4(normal, 1.0);
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:652]: I didn't think about this enough when I gave you this equation: mvpMatrix will transform object coordinates into our rendering coordinates.  But I don't think we actually want that for normals.  Specifically I think we want to apply rotations, but not translations.  Uniform scalings shouldn't matter, but non-uniform ones would.  So I think we need a separate matrix for applying to normals.  Assuming we're only scaling uniformly (x,y,z all the same scale-factor), then it's just rotation we'd want to use.  If I'm thinking about this right, in that case, an easy hack might be to just take "vec3 transformedZero = mvpMatrix * vec4(0.0, 0.0, 0.0, 1.0)" and then have "tmpNormal = tmpNormal - transformedZero".  Does that help improve normal results?

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 2]
```diff
@@ -626,36 +635,45 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
...
     // vertex position in threespace
     attribute vec4 position;
+    attribute vec3 normal;
 
     // texture coordinate for each vertex in normalized texture space (0..1)
     attribute mediump vec4 texture_coordinate;
 
     // texture coordinate for fragment shader (will be interpolated)
-    varying mediump vec2 sample_coordinate;
+    varying mediump vec3 sample_coordinate;
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:644]: We should revert the changes to sample_coordinate-- all our textures are 2D, so we don't want to be using 3D coordinates to index into them.  Same with below and in fragment shader.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 3]
```diff
@@ -626,36 +635,45 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
...
   const GLchar *frag_src = R"(
     precision mediump float;
 
-    varying vec2 sample_coordinate;  // texture coordinate (0..1)
+    varying vec3 sample_coordinate;  // texture coordinate (0..1)
+    varying vec3 vNormal;
     uniform sampler2D texture;  // texture to shade with
 
+    // Define ambient lighting constants
+    const vec3 ambient = vec3(-0.1, -0.1, -0.1);
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:665]: Having a negative ambient value doesn't make sense.  In order to "dim" or "brighten" we should be using multiplication in any event.  So for now you can try something like "vec3 ambient = (0.6, 0.6, 0.6)", and then have "gl_FragColor = vec4(pixelRGB * ambient, 1.0)".  If ambient is always going to be white/gray/untinted light, then we could even just make it a float constant.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 4]
```diff
@@ -626,36 +635,45 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
...
+      vec4 pixel = texture2D(texture, sample_coordinate.xy);
       // If the alpha (background) value is near transparent, then discard the
       // pixel, this allows the rendering of transparent background GIFs
       if (pixel.a < 0.2) discard;
+      // Convert rgba to rgb
+      vec3 pixelRGB = pixel.rgb;
 
-      // Set the fragment color to the transformed pixel
-      gl_FragColor = pixel;
+      gl_FragColor = vec4(pixelRGB + ambient, 1.0);
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:676]: nit: this is small enough that unless we plan on doing more operations to it, we can just put pixel.rgb here and simplify the code a little

File: mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc [hunk 5]
```diff
@@ -268,7 +270,11 @@ const Matrix3f MatricesManagerCalculator::GenerateUserRotationMatrix(const float
         // with the direction of finger movement from the user (system dependent)
         Eigen::AngleAxisf(-rotation_radians, Eigen::Vector3f::UnitY()) *
         Eigen::AngleAxisf(0.0f, Eigen::Vector3f::UnitZ()) *
-        Eigen::AngleAxisf(0.0f, Eigen::Vector3f::UnitX());
+        // The (M_PI * 0.5) must be added in order to adjust the default
+        // rendering of the object (the object should appear in the upright
+        // orientation upon initial render of the scene - this is entirely
+        // dependent on the construction of the .obj file).
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc:276]: This means different obj files might not be able to be used here.  This is fine for now, as we can easily "fix" up the obj models themselves while we have so few of them.  But we should add a TODO to handle this the correct way-- the "initial orientation" (and possibly "initial size" and "initial positioning") of the model should be properties of each model/sticker, so the "initial rotation" would be handled separately from the user-rotation calculations, to be combined later into some sort of "overall rotation"

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 6]
```diff
@@ -27,6 +27,7 @@
 #include "mediapipe/graphs/object_detection_3d/calculators/camera_parameters.pb.h"
 #include "mediapipe/graphs/object_detection_3d/calculators/gl_animation_overlay_calculator.pb.h"
 #include "mediapipe/graphs/object_detection_3d/calculators/model_matrix.pb.h"
+#include <math.h>
```
Comment (jianingwei) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:30]: Can you move this to before the other includes? usually c lib includes, then absolute path includes.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 7]
```diff
@@ -291,6 +293,10 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
       LOG(ERROR) << "Failed to read vertices for frame " << frame_count_;
       return false;
     }
+
+    int vertices_len = sizeof(triangle_mesh.vertices)/sizeof(triangle_mesh.vertices[0]);
```
Comment (jianingwei) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:297]: nit: space around '/'
nit: const int

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc (hunk 3 above)
Comment (jianingwei) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:665]: Is this our light direction?

File: mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc [hunk 8]
```diff
@@ -62,7 +62,7 @@ namespace {
 //  ANCHORS - Anchor data with x,y,z coordinates (x,y are in [0.0-1.0] range for
 //    position on the device screen, while z is the scaling factor that changes
 //    in proportion to the distance from the tracked region) [REQUIRED]
-//  IMU_DATA - float[3] of [roll, pitch, yaw] of device [REQUIRED]
+//  IMU_MATRIX - float[9] of device rotation matrix [REQUIRED]
```
Comment (jianingwei) [inline mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc:65]: Can you comment on whether this is row major or column major?

File: mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc [hunk 9]
```diff
@@ -189,12 +188,15 @@ ::mediapipe::Status MatricesManagerCalculator::Process(CalculatorContext* cc) {
...
-  const float roll = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[2];
-  // We can pre-generate the IMU submatrix as this remains unchanged for all objects
-  const Matrix3f imu_rotation_submatrix = GenerateIMURotationSubmatrix(yaw, pitch, roll);
+  // Device IMU rotation submatrix
+  Matrix3f imu_rotation_submatrix;
+  int idx = 0;
+  for (int x = 0; x < 3; x++) {
+    for (int y = 0; y < 3; y++) {
+      // Rotation matrix must be transposed in order to get inverse of phone movement
+      imu_rotation_submatrix(y, x) = cc->Inputs().Tag(kIMUMatrixTag).Get<float[]>()[idx++];
```
Comment (jianingwei) [inline mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc:197]: This matrix is column major. Is the input matrix also column major matrix?

File: mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc [hunk 10]
```diff
@@ -268,7 +270,11 @@ const Matrix3f MatricesManagerCalculator::GenerateUserRotationMatrix(const float
         // with the direction of finger movement from the user (system dependent)
         Eigen::AngleAxisf(-rotation_radians, Eigen::Vector3f::UnitY()) *
         Eigen::AngleAxisf(0.0f, Eigen::Vector3f::UnitZ()) *
-        Eigen::AngleAxisf(0.0f, Eigen::Vector3f::UnitX());
+        // The (M_PI * 0.5) must be added in order to adjust the default
+        // rendering of the object (the object should appear in the upright
+        // orientation upon initial render of the scene - this is entirely
+        // dependent on the construction of the .obj file).
+        Eigen::AngleAxisf(M_PI * 0.5f, Eigen::Vector3f::UnitX());
```
Comment (jianingwei) [inline mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc:277]: If this varies per object, should this be defined elsewhere?

File: mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc [hunk 11]
```diff
@@ -35,7 +35,7 @@ namespace {
   using Matrix3f = Eigen::Matrix3f;
   using DiagonalMatrix3f = Eigen::DiagonalMatrix<float, 3>;
   constexpr char kAnchorsTag[] = "ANCHORS";
-  constexpr char kIMUDataTag[] = "IMU_DATA";
+  constexpr char kIMUMatrixTag[] = "IMU_MATRIX";
```
Comment (jianingwei) [inline mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc:38]: suggest naming this to ROTATION_MATRIX.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc (hunk 3 above)
Comment (viksharma-mtv) [reply mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc]: This is our ambient (unchanging) lighting for the scene that is applied to every object. The actual lighting (spotlight effects) and the use of normals to create shading on the object is being done on a separate branch and will be part of PR #5.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc (hunk 1 above)
Comment (viksharma-mtv) [reply mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:651]: I have tested this out with Lambertian shading, and it appears to perform slightly better. I am working on the lighting in a separate branch and will report back with my findings in regard to the improvements/issues when using these new normal transformations with the Lullaby shaders code.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc (hunk 3 above)
Comment (jianingwei) [reply mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc]: What does the three values of 0.1 represent? Can you document them and explain how you decide those numbers? In general, constants beyond 0 and 1 would benefit from some documentation or use a named variable whose name is self explanatory.

File: mediapipe/graphs/instantmotiontracking/subgraphs/calculators/tracked_anchor_manager_calculator.cc [hunk 12]
```diff
@@ -96,70 +97,70 @@ REGISTER_CALCULATOR(TrackedAnchorManagerCalculator);
...
       }
     }
-    if(!anchor_is_being_tracked) {
-      // TODO: Pass in vector of boxes to prevent breaking of terms from SetOffset
-      cc->Outputs().Tag(kCancelTag).AddPacket(MakePacket<int>(box.id()).At(timestamp++));
+    // Check if initial anchor position has changed
+    bool anchor_replaced = false;
+    for(const Anchor &previous_anchor : previous_iteration_anchor_data_) {
+      if(previous_anchor.sticker_id == anchor.sticker_id) {
+        if(previous_anchor.x != anchor.x || previous_anchor.y != anchor.y) {
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/subgraphs/calculators/tracked_anchor_manager_calculator.cc:118]: So if the sticker_id *doesn't* match, then we don't replace the anchor, regardless of previous_anchor.x/y?  That seems unintuitive-- if this is intentional, we should have some comments for explanation.

(Also, if somehow the user manages to choose the same precise x/y location-- maybe from the Android app having some default setting-- then the anchor .z value wouldn't be updated.  Probably not a concern given how unlikely that is to occur with our current app setup, but I think from an architectural standpoint, it's generally nicer to have things be more event-driven or signal-driven, and not have to code special behavior into the values themselves).

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 13]
```diff
@@ -294,7 +294,98 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
+
+    // Compute every triangle surface normal and store them for later use.
+    for (int i = 0; i < len; i += 3) {
+      // Vector A = V2 - V1
+      float Ax, Ay, Az;
+      // Vector B = V3 - V1;
+      float Bx, By, Bz;
+      // Calculate V1 (Previous Vertex), V2 (Current Vertex), V3 (Next Vertex)
+      // Last vertex (Edge case)
+      if(i >= len-3) {
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:311]: We don't need 3 copies of this code here.  We could either refactor the calculations into their own function, or better yet, we could just use the % operator to wrap around automatically.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 14]
```diff
@@ -294,7 +294,98 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
       return false;
     }
 
+    // Use the vertices in order to generate normals for shader usage.
     triangle_mesh.normals.reset(new float[lengths[0]]);
+    float vertex_normals[lengths[0]];
+    float triangle_surface_normals[lengths[0]];
+    int len = lengths[0];
+
+    // Compute every triangle surface normal and store them for later use.
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:303]: The vertices are not stored in any particular order, so this normal calculation code doesn't really work (some vertices will be used in only 1 triangle, some vertices may be used in 10 triangles, etc.).  The triangle_mesh.triangle_indices list contains the indices into the vertex list for forming the actual triangles of our model.  So any normals need to be calculated over those.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 15]
```diff
@@ -756,13 +728,18 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
...
-    const vec3 kLightColor = vec3(0.2);
+    const vec3 kLightColor = vec3(0.3);
     // Exponent for directional lighting that governs diffusion of surface light
     const float kExponent = 1.0;
+    // Define direction of lighting effect source
+    const vec3 lightDir = vec3(0.0, -1.0, -0.6);
+    // Hard-coded view direction
+    const vec3 viewDir = vec3(0.0, 0.0, -1.0);
 
+    // DirectionalLighting procedure from imported from Lullaby @ https://github.com/google/lullaby
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:742]: super-nit: extra "from"

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 16]
```diff
@@ -788,20 +765,17 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
 
     void main() {
       // Sample the texture, retrieving an rgba pixel value
-      vec4 pixel = texture2D(texture, sampleCoordinate.xy);
+      vec4 pixel = texture2D(texture, sampleCoordinate);
       // If the alpha (background) value is near transparent, then discard the
       // pixel, this allows the rendering of transparent background GIFs
       if (pixel.a < 0.2) discard;
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:771]: This won't necessarily be at all related to world-space.  The sampleCoordinate only tells us where in our texture-map we need to look to color our pixel.  These textures usually are very compressed, and can be arbitrarily laid out in memory (like the head of the dino can be right next to its feet in the texture), so we have to use something else to get a pixel position.

We could derive something for this ourselves, but GLES gives us "gl_FragCoord", which I think should suffice.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 17]
```diff
@@ -756,13 +728,18 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
     // generate ambient lighting of the scene on the object. Range is [0.0-1.0],
     // with the factor being proportional to the brightness of the lighting in the
     // scene being applied to the object
-    const float kAmbientLighting = 0.95;
+    const float kAmbientLighting = 0.9;
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:731]: nit: this is very high, so you'll probably want to lower this so you can see the effect of the lighting better.  Also, in particular we will want to have kAmbient + kLightColor <= 1.0, because otherwise we will be potentially overflowing the normal [0,1] color bounds, so we can see clamping behavior.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 18]
```diff
@@ -402,6 +309,67 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
       LOG(ERROR) << "Failed to read indices for frame " << frame_count_;
       return false;
     }
+
+    // Set triangle_mesh normals for shader usage
+    int len = lengths[2];
+    triangle_mesh.normals.reset(new float[len]);
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:315]: We might start by calculating one normal for each triangle, but eventually we need to give to the shader one normal for each vertex (unless we do something fancier like use a texture for our normals).

So we probably want this length to be lengths[0], while the temporary array "triangle_normals" below should stay lengths[2].

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 19]
```diff
@@ -402,6 +309,67 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
+    float triangle_normals[len];
+
+    // Compute every triangle surface normal and store them for later use.
+    for (int idx = 0; idx < len; idx += 3) {
+      int v1 = triangle_mesh.triangle_indices.get()[idx];
+      int v2 = triangle_mesh.triangle_indices.get()[idx + 1];
+      int v3 = triangle_mesh.triangle_indices.get()[idx + 2];
+      // Previous (V1) vertex X,Y,Z indices in triangle_mesh.vertices
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:323]: super-nit: These are just 3 (ordered) vertices, there's not really any notion of previous/current/next.  The calculations here look good though.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 20]
```diff
@@ -402,6 +309,67 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
+      float Bx = V3x - V1x;
+      float By = V3y - V1y;
+      float Bz = V3z - V1z;
+      // Calculate cross product
+      triangle_normals[idx] = Ay * Bz - Az * By;
+      triangle_normals[idx + 1] = Az * Bx - Ax * Bz;
+      triangle_normals[idx + 2] = Ax * By - Ay * Bx;
+    }
+
+    // Normalize each triangle surface normal
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:350]: nit: this works, but might as well just do the normalization in the same loop above for simplicity

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 21]
```diff
@@ -402,6 +309,67 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
+    for(int i = 0; i < len; i+=3) {
+      float product = 0.0;
+      product = product + triangle_normals[i] * triangle_normals[i];
+      product = product + triangle_normals[i+1] * triangle_normals[i+1];
+      product = product + triangle_normals[i+2] * triangle_normals[i+2];
+      float magnitude = sqrt(product);
+      triangle_normals[i] /= magnitude;
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:357]: [No action necessary] We can assume our assets are well-formed because we're creating them.  But in general, this would be a good thing to have a debug-mode-only check for (specifically that magnitude isn't 0 so we don't divide by 0).

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 22]
```diff
@@ -402,6 +309,67 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
+      triangle_normals[i+1] /= magnitude;
+      triangle_normals[i+2] /= magnitude;
+    }
+
+    // Combine all triangle normals connected to each vertex by adding the X,Y,Z
+    // value of each adjacent triangle surface normal to every vertex and then
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:363]: The list of triangles we have (and now our new list of triangle normals) doesn't have any ordering in it, in terms of where the triangles or the normals are placed.  So adjacency can't help us here.

For example, we can have vertex 13 used in only one triangle (let's say the vertices are 13, 17, and 22).  But then clearly the normal for *that* triangle doesn't need any averaging-- we should just use it for vertex 13 directly.  Conversely, let's say vertex 17 is used in 4 triangles: (13, 17, 22), (17, 1, 5), (17, 103, 22), and (100, 2, 17).  Then we'd want the normal for vertex 17 to make use of all 4 of the normals corresponding to those 4 triangles.

So one thing we can do is for each vertex maintain a sum of all the triangle normals it had a part in making, and then average all of those together.

File: mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc [hunk 23]
```diff
@@ -62,7 +62,7 @@ namespace {
 //  ANCHORS - Anchor data with x,y,z coordinates (x,y are in [0.0-1.0] range for
 //    position on the device screen, while z is the scaling factor that changes
 //    in proportion to the distance from the tracked region) [REQUIRED]
-//  IMU_DATA - float[3] of [roll, pitch, yaw] of device [REQUIRED]
+//  IMU_ROTATIONS - float[9] of row-major device rotation matrix [REQUIRED]
```
[the remark this answers is not in the document]
Comment (jianingwei) [reply mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc]: +1 on Tyler's comment. Or DEVICE_ROTATION

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 24]
```diff
@@ -342,32 +344,45 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
-      product = product + triangle_normals[i+2] * triangle_normals[i+2];
-      float magnitude = sqrt(product);
-      triangle_normals[i] /= magnitude;
-      triangle_normals[i+1] /= magnitude;
-      triangle_normals[i+2] /= magnitude;
+      float normal_x = Ay * Bz - Az * By;
+      float normal_y = Az * Bx - Ax * Bz;
+      float normal_z = Ax * By - Ay * Bx;
+      // Add connected normal to each associated vertex
+      // It is also necessary to increment each vertex denominator for averaging
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:351]: I think you need to normalize here first too.  Otherwise, when you average a vector of (1, 0, 0) and (0, 0, 1000), you'll get something really close to (0, 0, 1) instead of (1/rt(2), 0, 1/rt(2)).  So let's also pull normalize into a quick separate function so we can reuse easily.

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 25]
```diff
@@ -311,24 +311,26 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
-    int len = lengths[2];
-    triangle_mesh.normals.reset(new float[len]);
-    float triangle_normals[len];
+    triangle_mesh.normals.reset(new float[lengths[0]]);
 
-    // Compute every triangle surface normal and store them for later use.
-    for (int idx = 0; idx < len; idx += 3) {
+    // Used for storing the vertex normals prior to averaging
+    float vertex_normals_sum[lengths[0]];
+    float vertex_avg_denom[lengths[0]];
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:318]: nit: might as well have the denom just be an integer-- or can even call it "*_count" to signify we're just counting up

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 26]
```diff
@@ -342,32 +344,45 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
-      float magnitude = sqrt(product);
-      triangle_normals[i] /= magnitude;
-      triangle_normals[i+1] /= magnitude;
-      triangle_normals[i+2] /= magnitude;
+      float normal_x = Ay * Bz - Az * By;
+      float normal_y = Az * Bx - Ax * Bz;
+      float normal_z = Ax * By - Ay * Bx;
+      // Add connected normal to each associated vertex
+      // It is also necessary to increment each vertex denominator for averaging
+      vertex_normals_sum[v1 * 3] += normal_x;
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:352]: nit: let's add a quick for-loop here and make this a little cleaner.  If we really wanted to, we could do that for our v's as well (v[0], v[1], v[2]), but I think we can leave those as-is for now to make things clearer

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 27]
```diff
@@ -342,32 +344,45 @@ bool GlAnimationOverlayCalculator::LoadAnimationAndroid(
...
+      vertex_normals_sum[v1 * 3 + 1] += normal_y;
+      vertex_normals_sum[v1 * 3 + 2] += normal_z;
+      vertex_avg_denom[v1 * 3] += 1;
+      vertex_avg_denom[v1 * 3 + 1] += 1;
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:356]: super-nit: if integer counts, then "+= 1" --> "++"

File: mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc [hunk 28]
```diff
@@ -771,7 +786,7 @@ ::mediapipe::Status GlAnimationOverlayCalculator::GlSetup() {
       if (pixel.a < 0.2) discard;
 
       // Convert pixel to 3D for lighting procedure
-      vec3 pixelPos = vec3(sampleCoordinate, 1.0);
+      vec3 pixelPos = vec3(gl_FragCoord);
```
Comment (tyrmullen) [inline mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc:789]: nit: gl_FragCoord.xyz instead of vec3(gl_FragCoord) for clarity
